### PR TITLE
Bug/1541 plugin menu order

### PR DESCRIPTION
**Pull request by ivankruchkoff**
## Summary

Addresses issue #1541 

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @ivankruchkoff ! This looks great, just a few very minor tweaks. The main thing is the changes to the tests. Let me know if you have any questions 👍

File: includes/Core/Admin/Screens.php [hunk 1]
```diff
@@ -152,14 +152,15 @@ function() {
 		add_filter(
 			'menu_order',
 			function( array $menu_order ) {
-				$new_order = array();
-				foreach ( $menu_order as $index => $item ) {
-					if ( 'index.php' === $item || 0 === strpos( $item, self::PREFIX ) ) {
-						$new_order[] = $item;
-						unset( $menu_order[ $index ] );
-					}
+				// Move the sitekit-dashboard menu item to be one after the index.php item if it exists.
```
[the remark this answers is not in the document]
Comment (ivankruchkoff) [reply includes/Core/Admin/Screens.php]: Updated

File: includes/Core/Admin/Screens.php [hunk 2]
```diff
@@ -152,14 +152,15 @@ function() {
...
+				$dashboard_index = array_search( 'index.php', $menu_order, true );
+				$sitekit_index   = array_search( self::PREFIX . 'dashboard', $menu_order, true );
+				if ( false === $dashboard_index || false === $sitekit_index ) {
+					return $menu_order;
 				}
-				return array_values( array_merge( $new_order, $menu_order ) );
+				unset( $menu_order[ $sitekit_index ] );
+				array_splice( $menu_order, ++$dashboard_index, 0, self::PREFIX . 'dashboard' );
```
[the remark this answers is not in the document]
Comment (ivankruchkoff) [reply includes/Core/Admin/Screens.php]: Updated the index

File: tests/phpunit/integration/Core/Admin/ScreensTest.php [hunk 3]
```diff
@@ -108,11 +108,19 @@ public function test_menu_order() {
...
+			$menu_order
+		);
+
 		$this->screens->register();
 
 		// Imitate WordPress core running these filters.
 		if ( apply_filters( 'custom_menu_order', false ) ) {
-			$menu_order = apply_filters( 'menu_order', $menu_order );
+			$menu_order            = apply_filters( 'menu_order', $menu_order );
+			$overridden_menu_order = apply_filters( 'menu_order', $overridden_menu_order );
```
[the remark this answers is not in the document]
Comment (ivankruchkoff) [reply tests/phpunit/integration/Core/Admin/ScreensTest.php]: Added a data provider, added the third case, and gave a one liner for each test case in the data provider.

**Comment by ivankruchkoff** [comment]
Addressed all CR feedback.

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @ivankruchkoff – this looks great 👍 

There are a few comments below about the data provider which would be good for you to review. They're so small that it doesn't make sense to send back here though so I'll apply those changes and we can get this merged 😄

**Comment by aaemnnosttv** [comment]
@googlebot I consent.

**Review by aaemnnosttv — DISMISSED** [review]
Awesome!

File: includes/Core/Admin/Screens.php [hunk 4]
```diff
@@ -152,14 +152,15 @@ function() {
...
 			function( array $menu_order ) {
-				$new_order = array();
-				foreach ( $menu_order as $index => $item ) {
-					if ( 'index.php' === $item || 0 === strpos( $item, self::PREFIX ) ) {
-						$new_order[] = $item;
-						unset( $menu_order[ $index ] );
-					}
+				// Move the Site Kit dashboard menu item to be one after the index.php item if it exists.
+				$dashboard_index = array_search( 'index.php', $menu_order, true );
+				$sitekit_index   = array_search( self::PREFIX . 'dashboard', $menu_order, true );
```
Comment (felixarntz) [inline includes/Core/Admin/Screens.php:157]: I think we cannot rely on `self::PREFIX . 'dashboard'` here - it was only using `self::PREFIX` before for a reason, which is in certain cases the Site Kit menu is not available under the `dashboard` slug but the `dashboard-splash` slug.

We may need to change this to a more dynamic search which finds the entry that starts with `self::PREFIX` instead.

File: tests/phpunit/integration/Core/Admin/ScreensTest.php [hunk 5]
```diff
@@ -99,30 +99,73 @@ public function test_removal_of_admin_notices_outside_sitekit( $hookname ) {
...
 
-	public function test_menu_order() {
-		$menu_order = array(
-			'index.php',
-			'third-party-plugin',
-			'edit.php',
-			'options-general.php',
-			'googlesitekit-dashboard',
+
+	public function menu_order_provider() {
```
Comment (felixarntz) [inline tests/phpunit/integration/Core/Admin/ScreensTest.php:103]: Let's add a test case here that uses another `googlesitekit-`-prefixed value, e.g. `googlesitekit-dashboard-splash`.

Comment (felixarntz) [inline tests/phpunit/integration/Core/Admin/ScreensTest.php:103]: Nit-picking, but WP core (and also most of Site Kit) always names the data provider methods `data_*` similarly to the `test_*` method name (i.e. in this case `data_menu_order`).

**Review by felixarntz — CHANGES_REQUESTED** [review]
The overall approach LGTM, but there's one thing we need to maintain from the way it was before.

File: tests/phpunit/integration/Core/Admin/ScreensTest.php (hunk 5 above)
Comment (aaemnnosttv) [reply tests/phpunit/integration/Core/Admin/ScreensTest.php:103]: I noticed that too (although they have many providers that don't use this convention as well), and we have both in our codebase (I'm pretty sure I've always used `_provider` as a suffix) but I'm happy to standardize on something.

Comment (ivankruchkoff) [reply tests/phpunit/integration/Core/Admin/ScreensTest.php:103]: Looked into it, we're at **31 dataProviders**. **17** with **_provider**, **14** with **data_** and this.
```
[/google-site-kit/tests/phpunit/integration]$ ack dataProvider|grep -v vendor|wc -l
      31
[/google-site-kit/tests/phpunit/integration]$ $ ack dataProvider|grep -v vendor|grep _provider |wc -l
      17
[/google-site-kit/tests/phpunit/integration]$ $ ack dataProvider|grep -v vendor|grep data_ |wc -l
      14
```

File: includes/Core/Admin/Screens.php (hunk 4 above)
Comment (ivankruchkoff) [reply includes/Core/Admin/Screens.php]: Updated the search to check for self::PREFIX

File: tests/phpunit/integration/Core/Admin/ScreensTest.php (hunk 5 above)
Comment (ivankruchkoff) [reply tests/phpunit/integration/Core/Admin/ScreensTest.php]: Added the fourth test case.

Comment (ivankruchkoff) [reply tests/phpunit/integration/Core/Admin/ScreensTest.php]: Updated to `data_menu_order`.

File: includes/Core/Admin/Screens.php [hunk 6]
```diff
@@ -152,14 +152,23 @@ function() {
...
-				foreach ( $menu_order as $index => $item ) {
-					if ( 'index.php' === $item || 0 === strpos( $item, self::PREFIX ) ) {
-						$new_order[] = $item;
-						unset( $menu_order[ $index ] );
+				// Move the Site Kit dashboard menu item to be one after the index.php item if it exists.
+				$dashboard_index = array_search( 'index.php', $menu_order, true );
+
+				$sitekit_index = false;
+				foreach ( $menu_order as $key => $value ) {
+					if ( substr( $value, 0, strlen( self::PREFIX ) ) === self::PREFIX ) {
```
Comment (felixarntz) [inline includes/Core/Admin/Screens.php:160]: Maybe simpler would be:
```suggestion
					if ( strpos( $value, self::PREFIX ) === 0 ) {
```

File: includes/Core/Admin/Screens.php [hunk 7]
```diff
@@ -152,14 +152,23 @@ function() {
...
+						$sitekit_index = $key;
+						$sitekit_value = $value;
```
Comment (felixarntz) [inline includes/Core/Admin/Screens.php:162]: I think we should only find the first item here, similar to how `array_search` would do it.
```suggestion
						$sitekit_value = $value;
						break;
```

**Review by felixarntz — APPROVED** [review]
Awesome!

**Pull request merged**